diff --git a/test/java/org/hibernate/search/test/configuration/LobTest.java b/test/java/org/hibernate/search/test/configuration/LobTest.java
index 1820470ca3..3ac74481e4 100644
--- a/test/java/org/hibernate/search/test/configuration/LobTest.java
+++ b/test/java/org/hibernate/search/test/configuration/LobTest.java
@@ -17,8 +17,7 @@
 
 import org.hibernate.Session;
 import org.hibernate.Transaction;
-import org.hibernate.dialect.Sybase11Dialect;
-import org.hibernate.dialect.SybaseASE15Dialect;
+import org.hibernate.dialect.SybaseDialect;
 import org.hibernate.search.FullTextQuery;
 import org.hibernate.search.FullTextSession;
 import org.hibernate.search.Search;
@@ -43,7 +42,7 @@
 public class LobTest extends SearchTestBase {
 
 	@Test
-	@SkipForDialect(value = { SybaseASE15Dialect.class, Sybase11Dialect.class },
+	@SkipForDialect(value = SybaseDialect.class,
 			comment = "Sybase does not support @Lob")
 	public void testCreateIndexSearchEntityWithLobField() {
 		// create and index
diff --git a/test/java/org/hibernate/search/test/envers/SearchAndEnversIntegrationTest.java b/test/java/org/hibernate/search/test/envers/SearchAndEnversIntegrationTest.java
index 3c55a59d0c..1c1c483475 100644
--- a/test/java/org/hibernate/search/test/envers/SearchAndEnversIntegrationTest.java
+++ b/test/java/org/hibernate/search/test/envers/SearchAndEnversIntegrationTest.java
@@ -12,7 +12,7 @@
 import java.util.List;
 
 import org.hibernate.Transaction;
-import org.hibernate.dialect.PostgreSQL81Dialect;
+import org.hibernate.dialect.PostgreSQLDialect;
 import org.hibernate.envers.AuditReader;
 import org.hibernate.envers.AuditReaderFactory;
 import org.hibernate.envers.query.AuditEntity;
@@ -38,7 +38,7 @@
  *
  * @author Davide Di Somma <davide.disomma@gmail.com>
  */
-@SkipForDialect(jiraKey = "HSEARCH-1943", value = PostgreSQL81Dialect.class)
+@SkipForDialect(jiraKey = "HSEARCH-1943", value = PostgreSQLDialect.class)
 @Category(PortedToSearch6.class)
 public class SearchAndEnversIntegrationTest extends SearchTestBase {
 
diff --git a/test/java/org/hibernate/search/test/query/LuceneQueryTest.java b/test/java/org/hibernate/search/test/query/LuceneQueryTest.java
index 004545c638..d4c4910561 100644
--- a/test/java/org/hibernate/search/test/query/LuceneQueryTest.java
+++ b/test/java/org/hibernate/search/test/query/LuceneQueryTest.java
@@ -237,7 +237,7 @@ public void testDefaultFetchSize() throws Exception {
 
 		ScrollableResults projections = hibQuery.scroll();
 		projections.beforeFirst();
-		Object[] projection = projections.get();
+		Object[] projection = (Object[]) projections.get();
 		assertNull( projection );
 
 		projections.next();
@@ -263,7 +263,7 @@ public void testFetchSizeLargerThanHits() throws Exception {
 		ScrollableResults results = hibQuery.scroll();
 		results.beforeFirst();
 		results.next();
-		Object[] result = results.get();
+		Object[] result = (Object[]) results.get();
 		assertEquals( "incorrect entityInfo returned", 1000, result[0] );
 
 		tx.commit();
@@ -286,14 +286,14 @@ public void testFetchSizeDefaultMax() throws Exception {
 		ScrollableResults results = hibQuery.scroll();
 		results.beforeFirst();
 		results.next();
-		Object[] result = results.get();
+		Object[] result = (Object[]) results.get();
 		assertEquals( "incorrect entityInfo returned", 1000, result[0] );
 		results.scroll( 2 );
-		result = results.get();
+		result = (Object[]) results.get();
 		assertEquals( "incorrect entityInfo returned", 1003, result[0] );
 		// check cache addition
 		results.next();
-		result = results.get();
+		result = (Object[]) results.get();
 		assertEquals( "incorrect entityInfo returned", 1004, result[0] );
 
 		tx.commit();
@@ -317,19 +317,19 @@ public void testFetchSizeNonDefaultMax() throws Exception {
 		ScrollableResults results = hibQuery.scroll();
 		results.beforeFirst();
 		results.next();
-		Object[] result = results.get();
+		Object[] result = (Object[]) results.get();
 		assertEquals( "incorrect entityInfo returned", 1000, result[0] );
 
 		results.next();
-		result = results.get();
+		result = (Object[]) results.get();
 		assertEquals( "incorrect entityInfo returned", 1002, result[0] );
 
 		results.scroll( 2 );
-		result = results.get();
+		result = (Object[]) results.get();
 		assertEquals( "incorrect entityInfo returned", 1004, result[0] );
 
 		results.next();
-		result = results.get();
+		result = (Object[]) results.get();
 		assertNull( result );
 
 		results.close();
@@ -338,12 +338,12 @@ public void testFetchSizeNonDefaultMax() throws Exception {
 
 		results.beforeFirst();
 		results.next();
-		result = results.get();
+		result = (Object[]) results.get();
 		assertEquals( "incorrect entityInfo returned", 1000, result[0] );
 
 		// And test a bad forward scroll.
 		results.scroll( 10 );
-		result = results.get();
+		result = (Object[]) results.get();
 		assertNull( result );
 
 		tx.commit();
@@ -364,7 +364,7 @@ public void testFetchSizeNonDefaultMaxNoHits() throws Exception {
 
 		ScrollableResults results = hibQuery.scroll();
 		results.beforeFirst();
-		Object[] result = results.get();
+		Object[] result = (Object[]) results.get();
 		assertNull( "non-null entity infos returned", result );
 
 		tx.commit();
@@ -427,11 +427,11 @@ public void testCurrent() throws Exception {
 
 		results = hibQuery.scroll();
 		results.scroll( 4 );
-		Object[] result = results.get();
+		Object[] result = (Object[]) results.get();
 		assertEquals( 1004, result[0] );
 
 		results.last();
-		result = results.get();
+		result = (Object[]) results.get();
 		assertEquals( 1004, result[0] );
 
 		tx.commit();
@@ -587,7 +587,7 @@ public void testScrollEmptyHits() throws Exception {
 		ScrollableResults projections = hibQuery.scroll();
 		projections.beforeFirst();
 		projections.next();
-		Object[] projection = projections.get();
+		Object[] projection = (Object[]) projections.get();
 		assertNull( projection );
 
 		hibQuery = fullTextSession.createFullTextQuery( query, Employee.class ).setMaxResults( 20 );
@@ -595,7 +595,7 @@ public void testScrollEmptyHits() throws Exception {
 		projections = hibQuery.scroll();
 		projections.beforeFirst();
 		projections.next();
-		projection = projections.get();
+		projection = (Object[]) projections.get();
 		assertNull( projection );
 
 		tx.commit();
diff --git a/test/java/org/hibernate/search/test/query/ProjectionQueryTest.java b/test/java/org/hibernate/search/test/query/ProjectionQueryTest.java
index 7cebc52a54..938872ca52 100644
--- a/test/java/org/hibernate/search/test/query/ProjectionQueryTest.java
+++ b/test/java/org/hibernate/search/test/query/ProjectionQueryTest.java
@@ -154,17 +154,17 @@ public void testProjectionWithScroll() throws Exception {
 
 		projections.beforeFirst();
 		projections.next();
-		Object[] projection = projections.get();
+		Object[] projection = (Object[]) projections.get();
 		checkProjectionFirst( projection, s );
 		assertTrue( projections.isFirst() );
 
 		projections.last();
-		projection = projections.get();
+		projection = (Object[]) projections.get();
 		checkProjectionLast( projection, s );
 		assertTrue( projections.isLast() );
 
 		projections.next();
-		projection = projections.get();
+		projection = (Object[]) projections.get();
 		assertNull( projection );
 
 		projections.close();
@@ -172,11 +172,11 @@ public void testProjectionWithScroll() throws Exception {
 		projections = hibQuery.scroll();
 
 		projections.first();
-		projection = projections.get();
+		projection = (Object[]) projections.get();
 		checkProjectionFirst( projection, s );
 
 		projections.scroll( 2 );
-		projection = projections.get();
+		projection = (Object[]) projections.get();
 		checkProjection2( projection, s );
 
 		//cleanup
diff --git a/test/java/org/hibernate/search/test/query/ScrollableResultsTest.java b/test/java/org/hibernate/search/test/query/ScrollableResultsTest.java
index 5121cd9c48..3244879d99 100644
--- a/test/java/org/hibernate/search/test/query/ScrollableResultsTest.java
+++ b/test/java/org/hibernate/search/test/query/ScrollableResultsTest.java
@@ -84,7 +84,7 @@ public void testScrollingForward() {
 			position++;
 			int bookId = position;
 			assertEquals( position, scrollableResults.getRowNumber() );
-			AlternateBook book = (AlternateBook) scrollableResults.get()[0];
+			AlternateBook book = (AlternateBook) ( (Object[]) scrollableResults.get() )[0];
 			assertEquals( bookId, book.getId().intValue() );
 			assertEquals( "book about the number " + bookId, book.getSummary() );
 			assertTrue( sess.contains( book ) );
@@ -112,7 +112,7 @@ public void testResultsAreManaged() {
 		int position = -1;
 		while ( scrollableResults.next() ) {
 			position++;
-			AlternateBook book = (AlternateBook) scrollableResults.get()[0];
+			AlternateBook book = (AlternateBook) ( (Object[]) scrollableResults.get() )[0];
 			assertTrue( sess.contains( book ) );
 			// evict some entities:
 			if ( position % 3 == 0 ) {
@@ -159,7 +159,7 @@ public void testScrollProjectionAndManaged() {
 		int position = scrollableResults.getRowNumber();
 		while ( scrollableResults.next() ) {
 			position++;
-			Object[] objs = scrollableResults.get();
+			Object[] objs = (Object[]) scrollableResults.get();
 			assertEquals( Employee.class, objs[0] );
 			assertEquals( position, objs[1] );
 			assertTrue( objs[2] instanceof Employee );
diff --git a/test/java/org/hibernate/search/test/session/MassIndexUsingManualFlushTest.java b/test/java/org/hibernate/search/test/session/MassIndexUsingManualFlushTest.java
index 05199ca806..58c43d0d4f 100644
--- a/test/java/org/hibernate/search/test/session/MassIndexUsingManualFlushTest.java
+++ b/test/java/org/hibernate/search/test/session/MassIndexUsingManualFlushTest.java
@@ -61,7 +61,7 @@ public void execute(Connection connection) throws SQLException {
 		int index = 0;
 		while ( results.next() ) {
 			index++;
-			final Email o = (Email) results.get( 0 );
+			final Email o = (Email) results.get();
 			s.index( o );
 			if ( index % 5 == 0 ) {
 				s.flushToIndexes();
