diff --git a/test/java/org/hibernate/search/integrationtest/mapper/orm/coordination/outboxpolling/OutboxPollingDisabledIT.java b/test/java/org/hibernate/search/integrationtest/mapper/orm/coordination/outboxpolling/OutboxPollingDisabledIT.java
index ef59998f6f..e43f3b38e5 100644
--- a/test/java/org/hibernate/search/integrationtest/mapper/orm/coordination/outboxpolling/OutboxPollingDisabledIT.java
+++ b/test/java/org/hibernate/search/integrationtest/mapper/orm/coordination/outboxpolling/OutboxPollingDisabledIT.java
@@ -83,7 +83,7 @@ public List<?> setupParams() {
 
 	@Test
 	public void metamodel_onlyUserEntities() {
-		assertThat( setupHolder.sessionFactory().getMetamodel().getEntities() )
+		assertThat( setupHolder.sessionFactory().getJpaMetamodel().getEntities() )
 				.extracting( EntityType::getName )
 				.containsOnly( IndexedEntity.NAME );
 	}
diff --git a/test/java/org/hibernate/search/integrationtest/mapper/orm/coordination/outboxpolling/OutboxPollingStrategyPropertyValueIT.java b/test/java/org/hibernate/search/integrationtest/mapper/orm/coordination/outboxpolling/OutboxPollingStrategyPropertyValueIT.java
index 0dd65e1228..e23cc0cb5a 100644
--- a/test/java/org/hibernate/search/integrationtest/mapper/orm/coordination/outboxpolling/OutboxPollingStrategyPropertyValueIT.java
+++ b/test/java/org/hibernate/search/integrationtest/mapper/orm/coordination/outboxpolling/OutboxPollingStrategyPropertyValueIT.java
@@ -103,7 +103,7 @@ public void resetFilter() {
 
 	@Test
 	public void metamodel_userEntitiesAndOutboxEventAndAgent() {
-		assertThat( setupHolder.sessionFactory().getMetamodel().getEntities() )
+		assertThat( setupHolder.sessionFactory().getJpaMetamodel().getEntities() )
 				.<Class<?>>extracting( Type::getJavaType )
 				.containsExactlyInAnyOrder( IndexedEntity.class, OutboxEvent.class, Agent.class );
 	}
