diff --git a/test/java/org/hibernate/search/integrationtest/mapper/orm/coordination/outboxpolling/OutboxPollingDisabledIT.java b/test/java/org/hibernate/search/integrationtest/mapper/orm/coordination/outboxpolling/OutboxPollingDisabledIT.java
index 163f7c5f1d..d263811c50 100644
--- a/test/java/org/hibernate/search/integrationtest/mapper/orm/coordination/outboxpolling/OutboxPollingDisabledIT.java
+++ b/test/java/org/hibernate/search/integrationtest/mapper/orm/coordination/outboxpolling/OutboxPollingDisabledIT.java
@@ -84,7 +84,7 @@ public List<?> setupParams() {
 
 	@Test
 	public void metamodel_onlyUserEntities() {
-		assertThat( setupHolder.sessionFactory().getMetamodel().getEntities() )
+		assertThat( setupHolder.sessionFactory().getJpaMetamodel().getEntities() )
 				.extracting( EntityType::getName )
 				.containsOnly( IndexedEntity.NAME );
 	}
diff --git a/test/java/org/hibernate/search/integrationtest/mapper/orm/coordination/outboxpolling/OutboxPollingStrategyPropertyValueIT.java b/test/java/org/hibernate/search/integrationtest/mapper/orm/coordination/outboxpolling/OutboxPollingStrategyPropertyValueIT.java
index 9d2c3771bc..72806b93cd 100644
--- a/test/java/org/hibernate/search/integrationtest/mapper/orm/coordination/outboxpolling/OutboxPollingStrategyPropertyValueIT.java
+++ b/test/java/org/hibernate/search/integrationtest/mapper/orm/coordination/outboxpolling/OutboxPollingStrategyPropertyValueIT.java
@@ -104,7 +104,7 @@ public void resetFilter() {
 
 	@Test
 	public void metamodel_userEntitiesAndOutboxEventAndAgent() {
-		assertThat( setupHolder.sessionFactory().getMetamodel().getEntities() )
+		assertThat( setupHolder.sessionFactory().getJpaMetamodel().getEntities() )
 				.<Class<?>>extracting( Type::getJavaType )
 				.containsExactlyInAnyOrder( IndexedEntity.class, OutboxEvent.class, Agent.class );
 	}
