diff --git a/main/java/module-info.java b/main/java/module-info.java
index 3bd5c117f3..28074f5739 100644
--- a/main/java/module-info.java
+++ b/main/java/module-info.java
@@ -16,12 +16,6 @@
 	requires org.hibernate.search.mapper.orm;
 	requires org.hibernate.search.backend.elasticsearch;
 
-	// This should be re-exported transitively by org.hibernate.search.mapper.orm
-	// but currently isn't, because org.hibernate.search.mapper.orm
-	// is still an automatic module
-	requires org.hibernate.search.engine;
-	requires org.hibernate.search.mapper.pojo;
-
 	/*
 	 * This is necessary in order to use SessionFactory,
 	 * which extends "javax.naming.Referenceable".
