diff --git a/test/java/org/hibernate/search/integrationtest/mapper/orm/cdi/CdiBeanResolutionIT.java b/test/java/org/hibernate/search/integrationtest/mapper/orm/cdi/CdiBeanResolutionIT.java
index 2e6cb3ec25..404958a8dc 100644
--- a/test/java/org/hibernate/search/integrationtest/mapper/orm/cdi/CdiBeanResolutionIT.java
+++ b/test/java/org/hibernate/search/integrationtest/mapper/orm/cdi/CdiBeanResolutionIT.java
@@ -139,7 +139,7 @@ private <T> void doTest(ExpectedScope expectedScope, CounterKeys counterKeys, Be
 
 		try ( @SuppressWarnings("unused")
 		SessionFactory sessionFactory = ormSetupHelper.start()
-				.withProperty( AvailableSettings.CDI_BEAN_MANAGER, cdiContainer.getBeanManager() )
+				.withProperty( AvailableSettings.JAKARTA_CDI_BEAN_MANAGER, cdiContainer.getBeanManager() )
 				.setup( IndexedEntity.class ) ) {
 			backendMock.verifyExpectationsMet();
 
diff --git a/test/java/org/hibernate/search/integrationtest/mapper/orm/cdi/CdiExtendedBeanManagerBootstrapShutdownIT.java b/test/java/org/hibernate/search/integrationtest/mapper/orm/cdi/CdiExtendedBeanManagerBootstrapShutdownIT.java
index 3fec78524f..4f964fdf2f 100644
--- a/test/java/org/hibernate/search/integrationtest/mapper/orm/cdi/CdiExtendedBeanManagerBootstrapShutdownIT.java
+++ b/test/java/org/hibernate/search/integrationtest/mapper/orm/cdi/CdiExtendedBeanManagerBootstrapShutdownIT.java
@@ -82,7 +82,7 @@ public void successfulBoot() {
 
 		try ( @SuppressWarnings("unused")
 		SessionFactory sessionFactory = ormSetupHelper.start()
-				.withProperty( AvailableSettings.CDI_BEAN_MANAGER, extendedBeanManager )
+				.withProperty( AvailableSettings.JAKARTA_CDI_BEAN_MANAGER, extendedBeanManager )
 				.setup( IndexedEntity.class ) ) {
 			// Hibernate Search should not have booted yet.
 			backendMock.verifyExpectationsMet();
@@ -138,7 +138,7 @@ public void failedBoot() {
 
 		try ( @SuppressWarnings("unused")
 		SessionFactory sessionFactory = ormSetupHelper.start()
-				.withProperty( AvailableSettings.CDI_BEAN_MANAGER, extendedBeanManager )
+				.withProperty( AvailableSettings.JAKARTA_CDI_BEAN_MANAGER, extendedBeanManager )
 				.setup( IndexedEntity.class ) ) {
 			// Hibernate Search should not have booted yet.
 			backendMock.verifyExpectationsMet();
@@ -193,7 +193,7 @@ public void cancelledBoot() {
 
 		try ( @SuppressWarnings("unused")
 		SessionFactory sessionFactory = ormSetupHelper.start()
-				.withProperty( AvailableSettings.CDI_BEAN_MANAGER, extendedBeanManager )
+				.withProperty( AvailableSettings.JAKARTA_CDI_BEAN_MANAGER, extendedBeanManager )
 				.setup( IndexedEntity.class ) ) {
 			// Hibernate Search should not have booted yet.
 			backendMock.verifyExpectationsMet();
@@ -235,7 +235,7 @@ public void failedShutdown() {
 
 		try ( @SuppressWarnings("unused")
 		SessionFactory sessionFactory = ormSetupHelper.start()
-				.withProperty( AvailableSettings.CDI_BEAN_MANAGER, extendedBeanManager )
+				.withProperty( AvailableSettings.JAKARTA_CDI_BEAN_MANAGER, extendedBeanManager )
 				.setup( IndexedEntity.class ) ) {
 			// Hibernate Search should not have booted yet.
 			backendMock.verifyExpectationsMet();
