diff --git a/main/java/module-info.java b/main/java/module-info.java
index 11aee01366..47a5e2599e 100644
--- a/main/java/module-info.java
+++ b/main/java/module-info.java
@@ -17,12 +17,6 @@
 	requires org.hibernate.search.backend.elasticsearch;
 	requires org.hibernate.search.mapper.orm.coordination.outboxpolling;
 
-	// This should be re-exported transitively by org.hibernate.search.mapper.orm
-	// but currently isn't, because org.hibernate.search.mapper.orm
-	// is still an automatic module
-	requires org.hibernate.search.engine;
-	requires org.hibernate.search.mapper.pojo;
-
 	/*
 	 * This is necessary in order to use SessionFactory,
 	 * which extends "javax.naming.Referenceable".
@@ -32,4 +26,10 @@
 
 	// Since Avro is using log4j we need to explicitly require it to make things work:
 	requires org.apache.logging.log4j;
+
+	/*
+	 * This is necessary in order to put ByteBuddy in the modulepath and make module exports effective.
+	 * I do not know why ByteBuddy doesn't end up in the modulepath without this.
+	 */
+	requires net.bytebuddy;
 }
