diff --git a/test/java/org/hibernate/search/documentation/search/query/QueryDslIT.java b/test/java/org/hibernate/search/documentation/search/query/QueryDslIT.java
index 6185209396..b7736775f3 100644
--- a/test/java/org/hibernate/search/documentation/search/query/QueryDslIT.java
+++ b/test/java/org/hibernate/search/documentation/search/query/QueryDslIT.java
@@ -61,7 +61,7 @@ public class QueryDslIT {
 	@Before
 	public void setup() {
 		entityManagerFactory = setupHelper.start()
-				.withProperty( AvailableSettings.JPA_SHARED_CACHE_MODE, SharedCacheMode.ENABLE_SELECTIVE.name() )
+				.withProperty( AvailableSettings.JAKARTA_SHARED_CACHE_MODE, SharedCacheMode.ENABLE_SELECTIVE.name() )
 				.setup( Book.class, Manager.class, Associate.class );
 		initData();
 	}
